File: 2018/6/main.py
from operator import itemgetter

coordinates = set()

with open('input.txt') as f:
    for coords in f.readlines():
        x, y = map(int, coords.split(", "))
        coordinates.add((x, y))

def calc_distance(a:tuple, b:tuple):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

min_x = min(coordinates, key=itemgetter(0))[0]
max_x = max(coordinates, key=itemgetter(0))[0]
min_y = min(coordinates, key=itemgetter(1))[1]
max_y = max(coordinates, key=itemgetter(1))[1]

infinite_coords = set([
    min(coordinates, key=lambda x: calc_distance(x, (min_x, min_y))),
    min(coordinates, key=lambda x: calc_distance(x, (min_x, max_y))),
    min(coordinates, key=lambda x: calc_distance(x, (max_x, min_y))),
    min(coordinates, key=lambda x: calc_distance(x, (max_x, max_y))),
])

closet_count = {c: [] for c in coordinates}
answer_2 = 0
for square in [(x, y) for x in range(min_x, max_x) for y in range(min_y, max_y)]:
    _sum = 0
    distances = {}
    for coord in coordinates:
        distance = calc_distance(square, coord)
        distances.setdefault(distance, set()).add(coord)
        _sum += distance

    if _sum < 10000:
        answer_2 += 1

    closest_coords = distances[min(distances.keys())]
    if len(closest_coords) == 1:
        closet_count[closest_coords.pop()].append(square)

answer_1 = max(
    len(points) for coord, points in closet_count.items()
    if coord not in infinite_coords
)
print(answer_1)
print(answer_2)

File: 2018/6/main.hs
import Data.List.Split (splitOn)
import Data.Maybe (catMaybes)
import Data.List

data Coord = Coord {x:: Int, y:: Int} 
    deriving (Show, Ord, Eq)

toCoord :: String -> Coord
toCoord coordStr = Coord (coordLst !! 0) (coordLst !! 1)
    where 
        coordLst = map read (splitOn ", " coordStr)

calcBounds :: [Coord] -> [Int]
calcBounds coords = [minimum xs, maximum xs, minimum ys, maximum ys]
    where
        xs = map x coords
        ys = map y coords

closetCoord :: [Coord] -> Coord -> Maybe Coord
closetCoord coords square = if length closestCoords == 1 then Just $ fst $ head $ closestCoords else Nothing
    where
        closestCoords = filter ((==shortest) . snd) distances
        shortest      = minimum $ map snd distances
        distances     = [(coord, calcDistance square coord) | coord <- coords]

calcDistance :: Coord -> Coord -> Int
calcDistance a b = abs (x a - x b) + abs (y a - y b)

main = do
    input <- fmap lines $ readFile "input.txt"
    let coords = map toCoord input
    let bounds = calcBounds coords
    let inifi = catMaybes [closetCoord coords (Coord x y) | x <- [bounds!!0, bounds!!1], y <- [bounds!!2, bounds!!3]]
    let grid = [Coord x y | x <- [bounds!!0..bounds!!1], y <- [bounds!!2..bounds!!3]]

    let results = filter (\ x -> notElem x inifi) $ catMaybes $ map (closetCoord coords) grid
    let answer1 = last $ sort $ map length $ group $ sort $ results
    putStrLn $ show $ answer1

    let answer2 = length [square | square <- grid, let d = sum $ map (calcDistance square) coords, d < 10000]
    putStrLn $ show $ answer2
